### ISSUE TITLE
Experiment Plan: Add generic links field to experiment plan.

### ISSUE THREAD

**Issue by aislinnpearson** [Rothamsted-Ecoinformatics/farm_rothamsted#458]
Rather than only having links for the experiment plan and experiment file, please can we convert this to a general hyperlink field, equivalent to what is on the experiment and proposal entities.

In an ideal situation: can we name links in the same way that we name files? This would make it easier to see what the link is for.

If this is possible, ideally we'd do it for all the experiment module entities and manually or programmatically add the names for the experiment plans and experiment files.

**Comment by paul121** [comment]
Unfortunately we don't have an easy way to add names/labels to links.

This reminds me there's actually an interesting case right now where `file` uploads can have a name/label, but `image` uploads cannot. Some people have expressed interest in letting file/images have additional metadata too.. basically creating another entity type that wraps the file/image itself. This entity could have a dedicated `links` field to accomplish this use-case.

Regen Farmer's Mutual in particular has asked for a "Document Asset" type altogether. My feeling is that a dedicated asset type is overkill, it is just the easiest/most tangible way to enable these features.

Does this resonate with you @aislinnpearson ?

**Comment by aislinnpearson** [comment]
Hi @paul121 

For this specific issue, I think we then maybe keep those two fields as dedicated links (and see how they change once we implement the electronic research notebooks) and then perhaps just add an additional field for links? What do you think?

More broadly regarding your comment here:

> Regen Farmer's Mutual in particular has asked for a "Document Asset" type altogether. My feeling is that a dedicated asset type is overkill, it is just the easiest/most tangible way to enable these features.

Thinking out loud, I think most of our documents are already linked to an asset which describes them (e.g. a soil survey should have an observation associated with it, an experiment upload is always linked to a experiment plan, harvest records are associated with a harvest log). In essence we already have the entity type describing the documents... but I can see how we might want to standardise the naming of those documents and/or tag them to be searchable. One of the document management systems that I have seen which I do I really like in the electronic lab notebooks is a document library (essentially a place where all your files are stored). So if you are looking for a document, you can search this library and/or find the asset it is associated with. But do I want to add metadata to the file and/or link that extends beyond what is in the existing entity types... I'm not sure. Certainly a name would be nice for links, but other than that a searchable document library is probably more important. Especially if we could use it to extract geometries, convert them and share those files... but that seems like a whole different use case to me.

In short, I'm struggling to picture what metadata we'd need for a file or link that isn't already captured. Really the only things I can think of are link names, and the ability to search within documents in a document library.

**Comment by paul121** [comment]
> I think we then maybe keep those two fields as dedicated links (and see how they change once we implement the electronic research notebooks) and then perhaps just add an additional field for links? What do you think?

This sounds good for now! I'll update the issue title.

Thinking out loud on the rest as well...

> One of the document management systems that I have seen which I do I really like in the electronic lab notebooks is a document library (essentially a place where all your files are stored). So if you are looking for a document, you can search this library and/or find the asset it is associated with.

This is good to know and would be great to reference! Document library is a larger goal many have interest in..

> But do I want to add metadata to the file and/or link that extends beyond what is in the existing entity types... I'm not sure. Certainly a name would be nice for links, but other than that a searchable document library is probably more important.

The challenge is that we basically have 3 types of "documents" for the document library: file upload (file entity), image upload (file entity) and links. Unfortunately only the file uploads allow you to add an optional Name/description. Images and links do not. If a new entity could host this "document" concept then this "document" could be named and allow a single file/image/link to be added to it. Everything could be named and have additional metadata (as needed) added as well.

**Comment by paul121** [comment]
Ahh I was mistaken! Drupal does actually have a setting to provide a "Link Text". So this isn't too bad to add! I think we should still use this same approach: leave experiment plan & experiment file as dedicated fields, but we can another catch all links field for other links.

![Screenshot from 2023-06-19 15-28-25](https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/assets/3116995/261435c2-e0dd-4adc-9d59-cab872b71b1c)

**Comment by aislinnpearson** [comment]
Oh, great! Thanks @paul121
